src/ripple/shamap/impl/TaggedPointer.ipp

```diff
@@ -0,0 +1,647 @@
+    toAllocate = numAllocatedChildren(toAllocate);
+    if (toAllocate == oldNumAllocated)
+        return;
+
+    // allocate hashes and children, but do not run constructors
+    TaggedPointer newHashesAndChildren{RawAllocateTag{}, toAllocate};
+    SHAMapHash *newHashes, *oldHashes;
+    std::shared_ptr<SHAMapAbstractNode>*newChildren, *oldChildren;
+    std::uint8_t newNumAllocated;
+    // structured bindings can't be captured in c++ 17; use tie instead
+    std::tie(newNumAllocated, newHashes, newChildren) =
+        newHashesAndChildren.getHashesAndChildren();
+    std::tie(std::ignore, oldHashes, oldChildren) = getHashesAndChildren();
+
+    if (newNumAllocated == SHAMapInnerNode::branchFactor)
+    {
+        // new arrays are dense, old arrays are sparse
+        iterNonEmptyChildIndexes(isBranch, [&](auto branchNum, auto indexNum) {
+            new (&newHashes[branchNum]) SHAMapHash{oldHashes[indexNum]};
+            new (&newChildren[branchNum]) std::shared_ptr<SHAMapAbstractNode>{
+                std::move(oldChildren[indexNum])};
+        });
+        // Run the constructors for the remaining elements
+        for (int i = 0; i < SHAMapInnerNode::branchFactor; ++i)
+        {
+            if ((1 << i) & isBranch)
+                continue;
+            new (&newHashes[i]) SHAMapHash{};
+            new (&newChildren[i]) std::shared_ptr<SHAMapAbstractNode>{};
+        }
+    }
+    else
+    {
+        // new arrays are sparse, old arrays may be sparse or dense
+        int curCompressedIndex = 0;
+        iterNonEmptyChildIndexes(isBranch, [&](auto branchNum, auto indexNum) {
+            new (&newHashes[curCompressedIndex])
+                SHAMapHash{oldHashes[indexNum]};
+            new (&newChildren[curCompressedIndex])
+                std::shared_ptr<SHAMapAbstractNode>{
+                    std::move(oldChildren[indexNum])};
+            ++curCompressedIndex;
+        });
+        // Run the constructors for the remaining elements
+        for (int i = curCompressedIndex; i < newNumAllocated; ++i)
+        {
+            new (&newHashes[i]) SHAMapHash{};
+            new (&newChildren[i]) std::shared_ptr<SHAMapAbstractNode>{};
+        }
+    }
+
+    *this = std::move(newHashesAndChildren);
+}
+
+inline TaggedPointer::TaggedPointer(std::uint8_t numChildren)
+    : TaggedPointer(TaggedPointer::RawAllocateTag{}, numChildren)
+{
+    auto [numAllocated, hashes, children] = getHashesAndChildren();
+    for (std::size_t i = 0; i < numAllocated; ++i)
+    {
+        new (&hashes[i]) SHAMapHash{};
+        new (&children[i]) std::shared_ptr<SHAMapAbstractNode>{};
+    }
+}
+
+inline TaggedPointer::TaggedPointer(TaggedPointer&& other) : tp_{other.tp_}
+{
+    other.tp_ = 0;
+}
+
+inline TaggedPointer&
+TaggedPointer::operator=(TaggedPointer&& other)
+{
+    if (this == &other)
+        return *this;
+    destroyHashesAndChildren();
+    tp_ = other.tp_;
+    other.tp_ = 0;
+    return *this;
+}
+
+TaggedPointer::~TaggedPointer()
```

Comment:
Missing `inline`.

Reply:
Thanks! Fixed in `c338b8280c [fold] Add missing inline`